File: src/services/onetime-code.ts
import { Db, Collection } from 'mongodb';
import { MongoService } from '../base';
import { randomBytes } from 'crypto';

export interface OnetimeCodeServiceOptions {
    collectionName?: string;
    ttlLimit?: number;
    createCodeFn?: () => string;
}

export interface OnetimeCodeRecord {
    owner: string;
    type: string;
    code: string;
    createdAt: Date;
}

function createSecureCode(): string {
    return randomBytes(16).toString('hex');
}

export class OnetimeCodeService extends MongoService {
    private collection: Collection<OnetimeCodeRecord>;
    private ttlLimit: number;
    private createCodeFn: () => string;

    constructor(db: Db, options: OnetimeCodeServiceOptions = {}) {
        super(db);
        const {
            collectionName = 'onetime_code',
            ttlLimit = 60 * 10, // 10 minutes
            createCodeFn = createSecureCode
        } = options;
        
        this.collection = this.db.collection(collectionName);
        this.ttlLimit = ttlLimit;
        this.createCodeFn = createCodeFn;
    }

    async buildIndex(): Promise<void> {
        await this.collection.createIndex(
            [
                { owner: 1 },
                { type: 1 },
                { code: 1 }
            ],
            { unique: true }
        );
        await this.collection.createIndex(
            { createdAt: 1 },
            { expireAfterSeconds: this.ttlLimit }
        );
    }

    async createCollections(): Promise<void> {
        // No additional collections needed
    }

    async createCode(user: string, codeType: string): Promise<string> {
        const maxAttempts = 3;
        for (let i = 0; i < maxAttempts; i++) {
            const code = this.createCodeFn();
            try {
                await this.collection.insertOne({
                    owner: user,
                    type: codeType,
                    code: code,
                    createdAt: new Date()
                });
                return code;
            } catch (error) {
                if (i === maxAttempts - 1) {
                    throw new Error('Failed to generate unique code after multiple attempts');
                }
                continue;
            }
        }
        throw new Error('Failed to generate unique code after multiple attempts');
    }

    async verifyCode(user: string, codeType: string, code: string): Promise<boolean> {
        const record = await this.collection.findOne({
            owner: user,
            type: codeType,
            code: code
        });

        if (!record) {
            return false;
        }

        const now = new Date();
        const createdAt = record.createdAt;
        const timeDiff = now.getTime() - createdAt.getTime();
        
        return timeDiff <= this.ttlLimit * 1000;
    }

    async consumeCode(user: string, codeType: string, code: string): Promise<boolean> {
        const result = await this.collection.findOneAndDelete({
            owner: user,
            type: codeType,
            code: code
        });
        return result !== null;
    }
} 

File: pysource/one_time_code.py
import secrets
from datetime import datetime, timedelta

from pymongo.synchronous.collection import Collection

from atkins.base import MongoService
from typing import TypedDict, Optional


class UserRecord(TypedDict):
    owner: str
    type: str
    code: str
    createdAt: datetime


def create_secure_code():
    return secrets.token_hex(16)


class OnetimeCodeService(MongoService):

    def __init__(
            self,
            db,
            collection_name: str = 'ont_time_code',
            ttl_limit=60 * 10,  # 10 minutes,
            create_code_fn=create_secure_code
    ):
        super().__init__(db)
        self.collection_name = collection_name
        self.coll: Collection[UserRecord] = self.db[collection_name]
        self.ttl_limit = ttl_limit
        self.create_code_fn = create_code_fn

    def build_index(self, **kwargs):
        self.coll.create_index(
            [
                ("owner", 1),
                ("type", 1),
                ("code", 1)
            ],
            unique=True
        )
        self.coll.create_index(
            "createdAt",
            expireAfterSeconds=self.ttl_limit
        )

    def create_collections(self, **kwargs):
        pass

    def create_code(self, user: str, code_type: str) -> str:
        """
        Create a new one-time code for the user and type.
        Ensures the code is unique for this user and type combination.
        """
        max_attempts = 3
        for _ in range(max_attempts):
            code = self.create_code_fn()
            try:
                self.coll.insert_one({
                    "owner": user,
                    "type": code_type,
                    "code": code,
                    "createdAt": datetime.now()
                })
                return code
            except Exception:
                continue
        raise Exception("Failed to generate unique code after multiple attempts")

    def verify_code(self, user: str, code_type: str, code: str) -> bool:
        """Check if the code exists for the given user and type, without consuming it."""
        record = self.coll.find_one({
            "owner": user,
            "type": code_type,
            "code": code,
        })

        if not record:
            return False
        if datetime.now() - record['createdAt'] > timedelta(seconds=self.ttl_limit):
            return False
        return True

    def consume_code(self, user: str, code_type: str, code: str) -> bool:
        """Verify and consume the code by removing it atomically."""
        result = self.coll.find_one_and_delete({
            "owner": user,
            "type": code_type,
            "code": code,
        })
        return result is not None
